### DISCUSSION TITLE
关于 基于 G2 5.0 版本的 G2Plot v3！

### DISCUSSION THREAD

**Discussion by hustcc** [antvis/G2Plot#3758] [Show and tell]
升级的 [PR](https://github.com/antvis/G2Plot/pull/3722) 在这里，但是 hold 不合并，后续暂时不会基于 G2 v5 迭代新版本的 G2Plot，这里做一些说明。

### 为什么会有 G2Plot 库？

在过去几年中，大量的低代码项目出来，可视化是低代码搭建的必备的组件，但是当时在基于 G2 v4、v3 去封装的时候，会带来大量的冗余代码，以及一些单一图表功能的封装，且当时 G2 的开放能力没有办法做出 `图表` 的概念，只能是基于图形语法的基础概念组合出图形。

所以基于 G2 封装了 G2Plot，其中的每一个图表都是对应一个切实含义的图表，比如：折线图、条形图、柱形图、饼图等等，期望降低业务使用 AntV 图表的门槛。

### 为什么不迭代 G2Plot v3

主要有几个原因：

1. G2 v5 具备有很好的开放扩展能力，能够像搭积木一样，逐步组合出复杂的图表，也能通过组合的方式产出 `图表` 的概念，比如，基于 G2 开放能力，封装的 [g2-extensions](https://github.com/antvis/g2-extensions/tree/master/plot)。在这样的架构设计上，G2Plot 的内容，更多都是属于 G2 的插件，或者属于生态的一部分。
2. 本 PR 尝试基于 G2 v5 封装 G2Plot v3 代码的时候，发现整体代码估计在 4000 行左右，而过去 G2Plot v2 这些代码在 1.8 万行，可以理解 G2 v5 通过架构方式已经融入了大部分的能力。
3. 维护成本考虑，为了 4000 行代码的封装，我们需要单独维护官网 DEMO、文档、教程等，而这些内容又和 G2 及其类似，所以整体 ROI 非常低，因此这部分代码目前在新版本 Ant Design Charts 中。
4. 减少概念，G2 和 G2Plot 定位区别是什么？这个开发团队能清楚，但是用户很难理解清楚，这也是困扰用户选型的大问题。这个问题是 AntV 整体技术栈都存在的问题：品牌太多，项目太多，之间差异又不是那么明确。所以接下来会对 G2Plot、Graphin、XFlow、L7Plot 等中间层封装，逐步淡化他们品牌感，而是做成对应 G2、G6、X6、L7 的周边生态项目。

### 如果我们业务产品使用了 G2Plot v2，后续怎么办？

有几个方案：

1. 轻度场景，有一定灵活性，逐步迁移 G2 v5。
2. 轻度场景，简单的图表使用，逐步迁移使用 Ant Design Charts。
3. 重可视化场景，基于 G2 v5 的插件开发方案，封装自己业务需要的图表库，如何公司可行的，也可以开源出来，我们会[收录到官网](https://github.com/antvis/G2/discussions/5772)，让更多人受益，也可以让 G2Plot 的事情在社区处处开花。

从我做开源的角度来说，非常建议使用 3 方案，不仅可以应对业务对可视化诉求的变化，也是理解 G2 的架构设计思路，以及可视化的原理。欢迎和大家一起讨论！

**Comment by hewitt99** [comment]
请问“G2 v5 的插件开发方案”具体是什么？
指的是创建一个服务于业务的图表库，按需引入吗？

**Reply by hustcc** [reply]
差不多可以这么理解，G2 5.0 提供了更好的业务图表扩展方式，可以参考 https://github.com/antvis/g2-extensions/tree/master/plot 写法，这个项目中是扩展了一个 sunburst 旭日图。
